File: retrocpu_emu/src/main/feature/board/io_ports.ts
/**
 * 1階 IO ボードのポートマップ（簡易）
 * 根拠: MN1613_CPUボードメモリ_IOマップ.mdc
 *
 * 0000 RESET_VECTOR — リセット時 CPU が読む起動 IC
 * 0020〜0024 — ハンドシェイク（任意で bridge 接続）
 */

import type { CpuIoSignals } from "../cpu/mn1613/mn1613ioport";
import {
  setIoReadCallback,
  setIoWriteCallback,
} from "../cpu/mn1613/mn1613";
import {
  createHandshakeIoPortBridge,
  IO_PORT as HSHK_IO_PORT,
} from "./handshake/io_port_bridge";

/** IO:0000 — リセットベクタ（ワードアドレス） */
export const IO_PORT_RESET_VECTOR = 0x0000;

/** モニター入口（メモリマップ上の既定） */
export const MONITOR_ENTRY_WORD = 0x0200;

/** H 命令オペコード */
export const OPCODE_H = 0x2000;

let _resetVector = MONITOR_ENTRY_WORD;
let _handshakeBus: CpuIoSignals | null = null;
let _intCause = 0;
let _interruptBusy = 0;

/**
 * リセット時に CPU が読む起動アドレスを返す。
 * @returns ワードアドレス（既定はモニター入口 0x0200）
 */
export function getResetVector(): number {
  return _resetVector & 0xffff;
}

/** IO ボード側が RESET_VECTOR レジスタに書く（モニター展開後に 0x0200 を流す） */
export function setResetVector(wordAddr: number): void {
  _resetVector = wordAddr & 0xffff;
}

/**
 * ハンドシェイク信号バスを登録する（0x20〜0x24 の委譲先）。
 * @param bus 接続するバス。null で切り離す
 */
export function attachHandshakeBus(bus: CpuIoSignals | null): void {
  _handshakeBus = bus;
}

/**
 * 割り込み要因（IO:0021 Bit0-2）を IO ボード側から設定する。
 * ハンドシェイクバス接続時はバスへ、未接続時は内部ラッチへ書く。
 * @param cause INT_CAUSE_CODE の値（下位 3bit のみ有効）
 */
export function setIntCause(cause: number): void {
  _intCause = cause & 0x07;
  if (_handshakeBus) {
    _handshakeBus.INT_CAUSE = _intCause as CpuIoSignals["INT_CAUSE"];
  }
}

/**
 * 割り込み処理中フラグ（IO:0020 Bit0）の現在値を返す。
 * CPU の割り込みハンドラが立てている間は 1。
 * @returns 0 または 1
 */
export function getInterruptBusy(): number {
  return _handshakeBus ? _handshakeBus.INTERRUPT_BUSY : _interruptBusy;
}

/**
 * ハンドシェイク転送中（HSHK_ENA=1）かどうかを返す。
 * バス未接続時は常に false。
 * @returns 転送中なら true
 */
export function isHandshakeActive(): boolean {
  return _handshakeBus ? _handshakeBus.HSHK_ENA === 1 : false;
}

/**
 * CPU の RD/WT コールバックを IO ボードポートに接続する。
 * 既存のハンドシェイク bridge があれば 0x20〜0x24 を委譲する。
 */
export function attachIoBoardPorts(): void {
  const hshk = _handshakeBus
    ? createHandshakeIoPortBridge(_handshakeBus)
    : null;

  setIoReadCallback((port) => {
    const p = port & 0xffff;
    if (p === IO_PORT_RESET_VECTOR) {
      return _resetVector & 0xffff;
    }
    if (
      hshk &&
      (p === HSHK_IO_PORT.INTERRUPT_BUSY ||
        p === HSHK_IO_PORT.INT_CAUSE ||
        p === HSHK_IO_PORT.HSHK_CTRL ||
        p === HSHK_IO_PORT.HSHK_IN_DATA ||
        p === HSHK_IO_PORT.HSHK_OUT_DATA)
    ) {
      return hshk.read(p);
    }
    // バス未接続でも割り込み要因・処理中フラグは CPU から見える必要がある
    if (p === HSHK_IO_PORT.INTERRUPT_BUSY) return _interruptBusy;
    if (p === HSHK_IO_PORT.INT_CAUSE) return _intCause;
    return 0;
  });

  setIoWriteCallback((port, val) => {
    const p = port & 0xffff;
    if (p === IO_PORT_RESET_VECTOR) {
      _resetVector = val & 0xffff;
      return;
    }
    if (hshk) {
      hshk.write(p, val);
      return;
    }
    if (p === HSHK_IO_PORT.INTERRUPT_BUSY) {
      _interruptBusy = val & 1;
    }
  });
}

File: retrocpu_emu_rs/src/cpuboard/io_ports.rs
//! MN1613 CPU ボード IO ポートマップ
//! 根拠: MN1613_CPUボードメモリ_IOマップ.mdc / HandShake.mdc
//!
//! IO 0000 / 0020–0025 / 0030–0034 / 0036–0037

use crate::cpuboard::handshake::wires::{
	encode_int1_cause, HandshakeWires, INT1_CAUSE_ADDR_BREAK, INT1_CAUSE_STEP,
};
use crate::cpuboard::mn1613::{
	AddrBusAccess, AddrComparatorBank, StepBreakUnit, IO_PORT_BREAK_ADDR_HI, IO_PORT_BREAK_ADDR_LO,
	IO_PORT_BREAK_CTRL, IO_PORT_BREAK_HIT, IO_PORT_BREAK_PREV, IO_PORT_STEP_DELAY, IO_PORT_STEP_ENA,
};

/// IO:0000 — リセットベクタ（ワードアドレス）
pub const IO_PORT_RESET_VECTOR: u16 = 0x0000;

/// モニターのリセットベクタ表先頭（`g_reset_vector`）
pub const MONITOR_ENTRY_WORD: u32 = 0x0108;
/// IO:0 の値からの STR 語オフセット
pub const RESET_VECTOR_STR_OFF: u16 = 2;
/// IO:0 の値からの IC 語オフセット
pub const RESET_VECTOR_IC_OFF: u16 = 3;

/// 保留中の IRQ（レベルと要因パック）
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PendingIrq {
	/// MN1613 割り込みレベル（比較器／ステップは 1）
	pub level: u8,
	/// IO:0021 に載せる値（下位 3bit）
	pub cause: u8,
}

/// CPU ボード IO レジスタ群（比較器・ステップ・ハンドシェイク線を束ねる）
#[derive(Debug, Clone)]
pub struct IoPorts {
	reset_vector: u32,
	wires: HandshakeWires,
	comparators: AddrComparatorBank,
	step_break: StepBreakUnit,
	pending_irq: Option<PendingIrq>,
}

impl Default for IoPorts {
	fn default() -> Self {
		Self::new()
	}
}

impl IoPorts {
	/// モニター既定リセットベクタで初期化する。
	///
	/// # Returns
	/// - 初期化済みインスタンスを返します。
	pub fn new() -> Self {
		Self {
			reset_vector: MONITOR_ENTRY_WORD,
			wires: HandshakeWires::new(),
			comparators: AddrComparatorBank::new(),
			step_break: StepBreakUnit::new(),
			pending_irq: None,
		}
	}

	/// 比較器・ステップ・線・IRQ を初期化する（リセットベクタは維持）。
	pub fn reset_peripherals(&mut self) {
		self.wires.reset();
		self.comparators.reset();
		self.step_break.reset();
		self.pending_irq = None;
	}

	/// 全状態を初期値へ戻す（リセットベクタもモニター既定）。
	pub fn reset(&mut self) {
		self.reset_vector = MONITOR_ENTRY_WORD;
		self.reset_peripherals();
	}

	/// IO:0 が返すリセットベクタ表先頭（ワードアドレス下位 16bit）。
	///
	/// # Returns
	/// - 32bit 値を返します。
	pub fn reset_vector(&self) -> u32 {
		self.reset_vector
	}

	/// IO ボード側が RESET_VECTOR レジスタに書く。
	///
	/// # Arguments
	/// - `word_addr`: ワードアドレス（下位 16bit を保持）
	pub fn set_reset_vector(&mut self, word_addr: u32) {
		self.reset_vector = word_addr & 0xffff;
	}

	/// ハンドシェイク線への参照。
	///
	/// # Returns
	/// - ハンドシェイク信号線への参照を返します。
	pub fn wires(&self) -> &HandshakeWires {
		&self.wires
	}

	/// ハンドシェイク線への可変参照。
	///
	/// # Returns
	/// - ハンドシェイク信号線への可変参照を返します。
	pub fn wires_mut(&mut self) -> &mut HandshakeWires {
		&mut self.wires
	}

	/// アドレス比較器バンク。
	///
	/// # Returns
	/// - アドレス比較器バンクへの参照を返します。
	pub fn comparators(&self) -> &AddrComparatorBank {
		&self.comparators
	}

	/// アドレス比較器バンク（可変）。
	///
	/// # Returns
	/// - アドレス比較器バンクへの可変参照を返します。
	pub fn comparators_mut(&mut self) -> &mut AddrComparatorBank {
		&mut self.comparators
	}

	/// ステップ・ワンショット。
	///
	/// # Returns
	/// - ステップ実行ユニットへの参照を返します。
	pub fn step_break(&self) -> &StepBreakUnit {
		&self.step_break
	}

	/// ステップ・ワンショット（可変）。
	///
	/// # Returns
	/// - ステップ実行ユニットへの可変参照を返します。
	pub fn step_break_mut(&mut self) -> &mut StepBreakUnit {
		&mut self.step_break
	}

	/// 割り込み処理中フラグ（IO:0020 Bit0）。
	///
	/// # Returns
	/// - 8bit 値を返します。
	pub fn interrupt_busy(&self) -> u8 {
		self.wires.interrupt_busy & 1
	}

	/// 割り込み要因パック（IO:0021 下位 3bit）。
	///
	/// # Returns
	/// - 8bit 値を返します。
	pub fn int_cause(&self) -> u8 {
		self.wires.int_cause & 0x07
	}

	/// IO ボード側から割り込み要因を設定する。
	///
	/// # Arguments
	/// - `cause`: ポート値（下位 3bit）
	pub fn set_int_cause(&mut self, cause: u8) {
		self.wires.int_cause = cause & 0x07;
	}

	/// 保留 IRQ を取り出しクリアする。無ければ None。
	///
	/// # Returns
	/// - 保留中 IRQ があれば `Some`、なければ `None` を返します。
	pub fn take_pending_irq(&mut self) -> Option<PendingIrq> {
		self.pending_irq.take()
	}

	/// 保留 IRQ を覗く（クリアしない）。
	///
	/// # Returns
	/// - 保留中 IRQ があれば `Some`、なければ `None` を返します。
	pub fn peek_pending_irq(&self) -> Option<PendingIrq> {
		self.pending_irq
	}

	/// バスアクセスを比較器に渡し、ヒットなら INT1・CAUSE=0 を上げる。
	///
	/// # Arguments
	/// - `access`: MEM/IO・RD/WR
	///
	/// # Returns
	/// - ヒットしたスロット番号。ヒットなしなら `None`。
	pub fn probe_addr(&mut self, access: &AddrBusAccess) -> Option<usize> {
		let hit = self.comparators.probe(access)?;
		self.raise_addr_break_irq();
		Some(hit)
	}

	/// 命令フェッチをステップユニットへ渡し、ヒットなら INT1・CAUSE=1。
	///
	/// # Arguments
	/// - `word`: 命令語
	///
	/// # Returns
	/// - ステップ条件ヒット時は `true`。
	pub fn on_instruction_fetch(&mut self, word: u16) -> bool {
		if self.step_break.on_instruction_fetch(word) {
			self.raise_step_break_irq();
			true
		} else {
			false
		}
	}

	fn raise_addr_break_irq(&mut self) {
		self.set_int_cause(encode_int1_cause(INT1_CAUSE_ADDR_BREAK));
		self.pending_irq = Some(PendingIrq {
			level: 1,
			cause: encode_int1_cause(INT1_CAUSE_ADDR_BREAK),
		});
	}

	fn raise_step_break_irq(&mut self) {
		self.set_int_cause(encode_int1_cause(INT1_CAUSE_STEP));
		self.pending_irq = Some(PendingIrq {
			level: 1,
			cause: encode_int1_cause(INT1_CAUSE_STEP),
		});
	}

	/// CPU の IO リード。未マップは 0。
	///
	/// # Arguments
	/// - `port`: ポート番号
	///
	/// # Returns
	/// - 16bit 値を返します。
	pub fn read(&mut self, port: u16) -> u16 {
		let p = port & 0xffff;
		if p == IO_PORT_RESET_VECTOR {
			return (self.reset_vector & 0xffff) as u16;
		}
		if let Some(v) = self.comparators.read_port(p) {
			return v;
		}
		if let Some(v) = self.step_break.read_port(p) {
			return v;
		}
		if let Some(v) = self.wires.read_port(p) {
			return v;
		}
		0
	}

	/// CPU の IO ライト。
	///
	/// # Arguments
	/// - `port`: ポート番号
	/// - `val`: 16bit 値
	pub fn write(&mut self, port: u16, val: u16) {
		let p = port & 0xffff;
		if p == IO_PORT_RESET_VECTOR {
			self.reset_vector = u32::from(val & 0xffff);
			return;
		}
		if p == IO_PORT_BREAK_CTRL
			|| p == IO_PORT_BREAK_ADDR_LO
			|| p == IO_PORT_BREAK_ADDR_HI
			|| p == IO_PORT_BREAK_HIT
			|| p == IO_PORT_BREAK_PREV
		{
			self.comparators.write_port(p, val);
			return;
		}
		if p == IO_PORT_STEP_ENA || p == IO_PORT_STEP_DELAY {
			self.step_break.write_port(p, val);
			return;
		}
		let _ = self.wires.write_port(p, val);
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::cpuboard::handshake::IO_PORT_INT_CAUSE;
	use crate::cpuboard::mn1613::{encode_break_ctrl, AddrComparatorSlot, BREAK_RDWR_RD};

	#[test]
	fn reset_vector_port() {
		let mut io = IoPorts::new();
		assert_eq!(io.read(IO_PORT_RESET_VECTOR), 0x0108);
		io.write(IO_PORT_RESET_VECTOR, 0x0200);
		assert_eq!(io.reset_vector(), 0x0200);
	}

	#[test]
	fn probe_raises_irq1_cause0() {
		let mut io = IoPorts::new();
		io.comparators_mut().set_slot(
			0,
			AddrComparatorSlot {
				enabled: true,
				io: false,
				rdwr: BREAK_RDWR_RD,
				addr: 0x50,
			},
		);
		assert_eq!(io.probe_addr(&AddrBusAccess::read(0x50, false)), Some(0));
		let irq = io.take_pending_irq().unwrap();
		assert_eq!(irq.level, 1);
		assert_eq!(irq.cause, 0);
		assert_eq!(io.read(IO_PORT_INT_CAUSE), 0);
	}

	#[test]
	fn step_raises_irq1_cause1() {
		let mut io = IoPorts::new();
		io.write(IO_PORT_STEP_DELAY, 0);
		io.write(IO_PORT_STEP_ENA, 1);
		assert!(!io.on_instruction_fetch(0x1000)); // skip
		assert!(io.on_instruction_fetch(0x1001));
		let irq = io.take_pending_irq().unwrap();
		assert_eq!(irq.level, 1);
		assert_eq!(irq.cause, 1);
	}

	#[test]
	fn break_ctrl_via_io_write() {
		let mut io = IoPorts::new();
		io.write(IO_PORT_BREAK_ADDR_LO, 0x1234);
		io.write(
			IO_PORT_BREAK_CTRL,
			encode_break_ctrl(0, true, false, BREAK_RDWR_RD),
		);
		let s = io.comparators().get_slot(0).unwrap();
		assert!(s.enabled);
		assert_eq!(s.addr, 0x1234);
	}
}
